core/src/main/java/org/mule/runtime/core/internal/policy/CompositeSourcePolicy.java

```diff
@@ -165,30 +149,43 @@ public CompositeSourcePolicy(List<Policy> parameterizedPolicies,
    */
   @Override
-  protected Publisher<CoreEvent> applyNextOperation(Publisher<CoreEvent> eventPub) {
+  protected Publisher<CoreEvent> applyNextOperation(Publisher<CoreEvent> eventPub, Policy lastPolicy) {
+
     return from(eventPub)
         .transform(flowExecutionProcessor)
-        .map(flowExecutionResponse -> {
-          Map<String, Object> originalResponseParameters = commonPolicy
-              .getResponseParamsProcessor(flowExecutionResponse)
-              .getSuccessfulExecutionResponseParametersFunction()
-              .apply(flowExecutionResponse);
-
-          Message message = getParametersTransformer()
-              .map(parametersTransformer -> parametersTransformer
-                  .fromSuccessResponseParametersToMessage(originalResponseParameters))
-              .orElseGet(flowExecutionResponse::getMessage);
-
-          return InternalEvent.builder(flowExecutionResponse)
-              .message(message)
-              .addInternalParameter(POLICY_SOURCE_ORIGINAL_RESPONSE_PARAMETERS, originalResponseParameters)
-              .build();
-        })
-        .cast(CoreEvent.class)
-        .onErrorMap(MessagingException.class, this::mapErrorInInnerExecution);
+        .map(flowExecutionResponse -> policyEventMapper.onFlowFinish(flowExecutionResponse, getParametersTransformer()))
+        .onErrorContinue(MessagingException.class, (error, v) -> {
+          MessagingException me = (MessagingException) error;
+
+          if (isEventContextHandledByThisNext(me.getEvent())) {
+            lastPolicy.getPolicyChain().getNotificationHelper().fireNotification(me.getEvent(), me, AFTER_NEXT);
+
+            pushAfterNextFlowStackElement(lastPolicy).accept(me.getEvent());
+
+            me.setProcessedEvent(policyEventMapper.onFlowError(me.getEvent(), lastPolicy.getPolicyId(),
+                                                               getParametersTransformer()));
+
+            ((BaseEventContext) me.getEvent().getContext()).error(error);
+          }
+        });
+  }
+
+  private Consumer<CoreEvent> pushAfterNextFlowStackElement(Policy lastPolicy) {
+    return event -> ((DefaultFlowCallStack) event.getFlowCallStack())
+        .push(new FlowStackElement(toPolicyLocation(lastPolicy.getPolicyChain().getLocation()), null));
   }
 
-  private MessagingException mapErrorInInnerExecution(MessagingException messagingException) {
+  private String toPolicyLocation(ComponentLocation componentLocation) {
+    return componentLocation.getParts().get(0).getPartPath() + "/" + componentLocation.getParts().get(1).getPartPath()
+        + "[after next]";
+  }
+
+  private boolean isEventContextHandledByThisNext(CoreEvent event) {
+    final Set<String> eventCtxIds = ((InternalEvent) event).getInternalParameter(POLICY_NEXT_EVENT_CTX_IDS);
+    return eventCtxIds != null && eventCtxIds.contains(event.getContext().getId());
+  }
+
+  private InternalEvent mapErrorInInnerExecution(MessagingException messagingException) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unused private "mapErrorInInnerExecution" method. ([squid:UnusedPrivateMethod](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUnusedPrivateMethod))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7760&issues=AWnGD5CXotIXnmsUaS7k&open=AWnGD5CXotIXnmsUaS7k)

[ISSUE_KEY]: AWnGD5CXotIXnmsUaS7k